[codemod] Add jss to tss-react codemod

ryancogswell opened the pull request:
Initial naive implementation. Many more test cases left to add.

<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

ryancogswell commented:
I now have an implementation that handles some non-trivial `makeStyles` use cases. Here is the list of scenarios I want to support before taking this out of draft mode:

- [x] Transform imports for `withStyles` and `makeStyles` from `@material-ui/core/styles`, `@material-ui/core`, and  `@mui/styles`
- [x] Remove imports and usages of `createStyles`
- [x] Transform usages of `makeStyles` with an object argument
- [x] Transform usages of `makeStyles` with an arrow function argument with a single `theme` parameter
- [ ] Transform usages of `makeStyles` with an arrow function argument with no parameters
- [ ] Transform usages of `makeStyles` with an object argument with nested selectors
- [x] Transform usages of `makeStyles` with an arrow function argument with a single `theme` parameter with nested selectors
- [ ] Transform usages of `makeStyles` with an arrow function argument with no parameters with nested selectors
- [ ] Transform usages of `withStyles` with an object argument defined inline
- [ ] Transform usages of `withStyles` with an object argument. Object assigned to a top-level variable.
- [ ] Transform usages of `withStyles` with an arrow function argument with a single `theme` parameter. Arrow function assigned to a top-level variable.
- [ ] Transform usages of `withStyles` with an arrow function argument with no parameters. Arrow function assigned to a top-level variable.
- [ ] Transform usages of `withStyles` with an arrow function argument with a single `theme` parameter. Arrow function defined inline.
- [ ] Transform usages of `withStyles` wrapped by another HOC (shouldn't require any additional code).
- [ ] Transform usages of `withStyles` with an arrow function argument with a single `theme` parameter with nested selectors combined with pseudo-classes. Arrow function assigned to a top-level variable. Both js and ts.

ryancogswell commented:
@mnajdova I haven't verified this against my own code base yet, but it now covers all of the scenarios that I am aware of in my code base. The main features it doesn't handle are params to `useStyles` (we completely avoided dynamic styles in JSS in our code base) and providing a stylesheet name.

mnajdova commented:
> @mnajdova I haven't verified this against my own code base yet so I might have follow-up tweaks once I do, but it now covers all of the scenarios that I am aware of in my code base.

That's great! Should we wait for your verification in your project? I think it's healthy if we have tested it on at least one project.

> The main features it doesn't handle are params to useStyles (we completely avoided dynamic styles in JSS in our code base) and providing a stylesheet name.

Alright, I may spend some time this quarter on it, or we could try to find some help from the community of tss-react cc @garronej 

------------

All in all great effort @ryancogswell

ryancogswell commented:
> Should we wait for your verification in your project?

@mnajdova It was easy for me to see how to use it on a project after it is released. It was less clear to me how to use the codemod utility with an unreleased code mod. What’s the equivalent of doing 

```sh
 npx @mui/codemod v5.0.0/jss-to-tss-react <path>
 ```

using the code from my branch?

garronej commented:
Great effort @ryancogswell. Thank you!  
Hi @mnajdova,  
Thank you for following up.  
Two things that discouraged me from implementing this codemod and focus on documentation instead:  
1) TSS require a `cache` to be manually provided.  
2) Specific configurations are needed for making TSS work with SSR.  
Those two blockers would be cleared if [this effort](https://github.com/mui/material-ui/pull/32067#issuecomment-1086113700) goes through. I hope it does! 🤞🏻  

Best regards,

ryancogswell commented:
@garronej After reading through #32067, #28045, and doing some experimenting of my own, I'm realizing that there is more work I need to do on this codemod. I think I should have the codemod replace usages of `clsx` and `classnames` with `cx` when the usages are in the same scope as a `useStyles` call. I think it's the only way to provide developers with clear control over which styles should win. Unfortunately, the precedence of styles will **not** be the same as JSS. On a large project it will result in a fair amount of pain due to requiring manual, error-prone changes.

In JSS, stylesheet insertion order was controlled by the order of the calls to `makeStyles` and the styles for the different rules within a `makeStyles` call were kept in their respective order.

With tss-react, stylesheet insertion order usually matches the order of the calls to `useStyles`. For nested components, this means the styles are in the opposite order compared to JSS. `makeStyles` is called at the time of importing a component, so an inner component's `makeStyles` will be called before the outer component's `makeStyles`; whereas `useStyles` is called during render with the outer component's `useStyles` being called first.

**Outer.js**
```
import { makeStyles } from 'tss-react/mui';
// Inner `makeStyles` called during import before Outer `makeStyles`
import Inner from './Inner';

// Outer `makeStyles`
const useStyles = makeStyles()({ outerClass: { color: "red" } });

export default function Outer() {
   // Outer `useStyles` called before Inner `useStyles`
   const { classes } = useStyles();
   return <Inner className={classes.outerClass}/>;
}
```

**Inner.js**
```
import { makeStyles } from 'tss-react/mui';
import clsx from 'clsx';

const useStyles = makeStyles()({ innerClass: { color: "black" } });

export default function Inner({className}) {
   const { classes } = useStyles();
   return <div className={clsx(className, classes.innerClass)}/>;
}
```

For JSS, the styles in "outerClass" would have won over "innerClass", but with tss-react the precedence will be reversed. This can be fixed using `cx`, but this also requires checking/fixing the order of the classes passed to `cx`. For `clsx/classnames`, the order that classes are passed to it has no bearing on which styles win, so existing code will likely have classes in an order that does not match the precedence of the styles. For instance `Inner` in my example would need to be changed to `cx(classes.innerClass, className)` in order to get the styles to match the JSS behavior. It would be possible to do some fixing of the order in the codemod, though it will make the codemod a fair amount more complicated. It should also be possible to insert a comment in the cases where the order needs to be manually checked.

I have 750 usages of `clsx` in my application, so automating as much of the re-ordering as I can probably makes sense.

garronej commented:
Thank you for your sorrow investigation and for taking the time to write it down.  
I linked your message [in the documentation](https://docs.tss-react.dev/readme-1) for peoples that would want to acquire a better understanding of how JSS works compared to TSS.  

I think being able to control the ordering with `cx` is much preferable to having the priority defined by the order in which `makeStyles` are called. It's easier to reason about, it gives greater control and prevent refactoring from having unintended consequences on styles.  

So yes, `cx` is great, but it makes it harder to write a JSS to TSS codemod.

mnajdova commented:
@ryancogswell nice catch on the differences between JSS vs TSS insertion order (adding this in a note for the codemod description would definitely help the people migrating). 

Deciding on whether to add the support for the `clsx` to `cx` automation based on a real project usage sounds fair. We can try to release and see the feedback. To be fair, checking the `clsx` usages is less error prone (and easier?) than migrating the whole codebase from `makeStyles` to `styled()` + `sx`, so for developers that would not want to do it, it should be a win already.

@garronej I am adding https://github.com/mui/material-ui/pull/32067#issuecomment-1086113700 on my list of things for this week (will try to take a first look ASAP, probably today).

mnajdova commented:
> I think it will be ready for release once I add the clsx to cx replacement and add some documentation about the manual changes that may be necessary to maintain the style precedence that occurred with JSS. I plan to tackle both this week.

Alright, I am keeping an eye on the PR, feel free to mention me when we are closer to get a review :)

ryancogswell commented:
@mnajdova Running it against my own project caught a number of scenarios I wasn't previously handling sufficiently, so it's quite a bit more robust now. I found a few cases of using params in my app, so I went ahead and added support for that, and I added support for specifying the `name` option. I didn't add support for cases of using params with the arrow function specified at the CSS prop level -- only at the rule level. I'll add a test case tomorrow to make sure I add a TODO comment if the arrow function is specified at a deeper level (or I might add support for it if it doesn't complicate things much). I also have a few documentation updates left to do. I should be completely done (at least for now) with it ready to merge by midday (US Central Time) tomorrow.

garronej commented:
@ryancogswell Thank you for all your work, this is amazing.

ryancogswell commented:
@mnajdova I believe this is now ready to merge.

ryancogswell commented:
@mnajdova Please let me know if there’s anything I can do to help move the review of this forward.

ryancogswell commented:
@garronej There are 9 pairs of test files in the [`mui-codemod/src/v5.0.0/jss-to-tss-react.test` directory](https://github.com/mui/material-ui/tree/734d36281fd41aa766b7402bede79817b24c7fbd/packages/mui-codemod/src/v5.0.0/jss-to-tss-react.test). Could you please help with the review by looking through each pair (each `actual*` file has a corresponding `expected*` file showing what the codemod should produce) of test files to verify that the expected results match what you would expect/recommend.

You can also help by reviewing my changes to the documentation. In particular, I removed your `#### Overriding styles - classes prop` section. Instead, the [comprehensive example](https://github.com/mui/material-ui/blob/734d36281fd41aa766b7402bede79817b24c7fbd/docs/data/material/guides/migration-v4/migration-v4.md?plain=1#L2859) includes an example of how the codemod would transform use of the `classes` prop and includes a link to the `tss-react` documentation about it.

garronej commented:
@ryancogswell Yes yes, sorry, I have a lot on my plate but I will definitively have a look ASAP.

ryancogswell commented:
@garronej Most of the examples are pretty straightforward, and I've already verified that they all work (at least once all the MUI imports are also transformed by other codemods), so I think it will be pretty quick to look through.

garronej reviewed (approved):
You did an impressive job congratulation 👏🏻🎉  
I just added some comments about what, I think, are best practice but feel free to ignore.

ryancogswell commented on packages/mui-codemod/src/v5.0.0/jss-to-tss-react.test/expected-from-material-ui-core.js:
```diff
@@ -0,0 +1,49 @@
+}
+export default function ComponentUsingStyles(props) {
+  const { classes, cx } = useStyles(props, {
+    props: props
```
There isn't any guarantee that `props` will be called `props`, so I would need to make a special case to do this simplification when it is named `props` -- I'm not going to bother with it.

garronej replied:
Fair enough

ryancogswell commented on packages/mui-codemod/src/v5.0.0/jss-to-tss-react.test/expected-typescript-docs-example-params.tsx:
```diff
@@ -0,0 +1,43 @@
+  }
+}));
+
+function App({classes: classesProp}: {classes?: any}) {
+  const { classes, cx } = useStyles({
+    color: 'primary',
+    padding: 30
+  }, {
+    props: {
+      classes: classesProp
+    }
+  });
```
This isn't really about cutting corners. `classes` is being renamed to `classesProp` in the code that is the starting point for the codemod. That variable name could then be used beyond just in the `useStyles` call, so I can't mess with that.

garronej replied:
Fair point.

ryancogswell commented on packages/mui-codemod/src/v5.0.0/jss-to-tss-react.test/expected-typescript-docs-example-params.tsx:
```diff
@@ -0,0 +1,43 @@
+import { makeStyles } from 'tss-react/mui';
+
+const useStyles = makeStyles<{color: 'primary' | 'secondary', padding: number}, 'child' | 'small'>({name: 'App'})((theme, { color, padding }, classes) => ({
```
This part of the transformation is pretty unintelligent. If the original code is passing a second argument to `makeStyles`, then I move that entire "options" argument (without any further transformation) to be the argument to the `tss-react` `makeStyles` call. I'm not going to make this any more sophisticated.

ryancogswell commented:
@garronej Thanks for reviewing things! In looking over the migration guide I noticed a few broken links to your GitHub readme that I just fixed to go to the equivalent docs.tss-react.dev URL, so that will give people a resource for best practices in using tss-react. For now, I think I'll leave my "comprehensive example" as it is -- with the focus being to show what the codemod will do in a more complicated scenario rather than trying to duplicate the best-practices guidance that is in your documentation.

garronej commented:
Fair enough

ryancogswell commented:
@garronej And thanks for creating tss-react! Our company's application is migrated to v5 in our testing environment with the intention of moving v5 to production this weekend. We wouldn't be able to make this transition nearly as quickly or confidently if we had needed to move the styles (700 usages of `makeStyles`) to a different implementation approach and it's nice to be able to make a clean break from the JSS implementation so that our code can be StrictMode compatible going forward.

garronej commented:
Hi @ryancogswell,  
Just so you know there is now an ESLint plugin available for detecting unused classes:  
https://docs.tss-react.dev/detecting-unused-classes  
I want to document it but I'll wait for this PR to be merged first to avoir conflict.  

Best regards.

mnajdova commented:
I am going to do a review on this tomorrow, looks promising @ryancogswell :)

mnajdova commented on packages/mui-codemod/README.md:
````diff
@@ -605,6 +605,75 @@ You can find more details about this breaking change in [the migration guide](ht
+npx @mui/codemod v5.0.0/jss-to-tss-react <path>
+```
+
+You can find more details about migrating from JSS to tss-react in [the migration guide](https://mui.com/guides/migration-v4/#2-use-tss-react).
````
Would be great if we can list here things that are not supported, for example the "TODOs" added in `expected-todo-comments.js`. 

This will make sure we are transparent and will help future contributors to take features one by one and add them as part of the codemod.

ryancogswell replied:
I've added documentation about the scenarios that produce TODO comments.

mnajdova reviewed (commented):
I couldn't wait till tomorrow :D Just small details added, looks great overall.

mnajdova reviewed (approved):
Great job @ryancogswell! Can't wait to share this with the community :)

garronej commented:
Thank you very much @ryancogswell, you did a great job!
